src/librustc/infer/fudge.rs

```diff
@@ -124,6 +121,40 @@ impl<'a, 'gcx, 'tcx> TypeFolder<'gcx, 'tcx> for RegionFudger<'a, 'gcx, 'tcx> {
         self.infcx.tcx
     }
 
+    fn fold_ty(&mut self, ty: Ty<'tcx>) -> Ty<'tcx> {
+        match ty.sty {
+            ty::TyInfer(ty::InferTy::TyVar(vid)) => {
+                match self.type_variables.get(&vid) {
```

Comment:
Hmm, yes, I don't relish the thought of coercions happening or not happening based on very small changes. I was hoping that we could do away with this function and just sometimes unify the return type *before* the arguments, but that's not **precisely** equivalent to what it does today. (In particular, the would *force* the return type to be a subtype of the output type, whereas this generates an expected type based on the assumption that it *ought* to be so, but doesn't add a hard constraint.)

An obvious option is not to consider the "roots" of variables -- basically any appearance of some variable `?T` that was introduced during the snapshot would be mapped to a fresh variable outside the snapshot. I'm not sure just why I didn't do this; it seems like it *ought* to be harmless.

At the time when I wrote this code, I was nervous about "losing" the connection, but having thought more about it I think it should be safe for us to play more-or-less arbitrary games with the expected type without (at least) compromising soundness.  (This function mildly gives me the willies: e.g., it supplies an "expected-has-type" hint to the arguments, but that's stronger than is needed, etc.)

I think in my ideal world the "expected type" would be allowed to have holes (rather than fresh type variables per se), that have no effect during unification, and we would then translate **all** unbound type variables to these holes. But doing that would be a big refactor and of dubious (or maybe even negative) value anyway. 

Anyway, so long story short, what I could do is to try just replacing all newly created, unbound type variables with newly created, unbound type variables (ignoring the union-find effects). This would probably be backportable too, I imagine. We would lose the "subtype connection" in some cases for these variables, but that shouldn't matter, since this just affects the expected type, and the subtypes should be re-established when the ultimate type is produced.